### PR TITLE
Fix correctness issues in regexp and add `\r` and `\n` to fuzz tests

### PR DESCRIPTION
**Pull request by andygrove**
Signed-off-by: Andy Grove <andygrove@nvidia.com>

Closes https://github.com/NVIDIA/spark-rapids/issues/4521

Recent testing has exposed a number of compatibility issues related to end of string/line anchors and newlines. This PR fixes these correctness issues but also is a step backward in terms of what we support on the GPU. I am now working on re-implementing the features that have regressed to follow on from this PR, now that we have more comprehensive fuzz testing that now includes the `\r` and `\n` characters.

Changes in this PR:

- `\Z` and `$` now fall back to CPU, but `\z` is still supported on the GPU
- Fuzz tests now include `\r` and `\n`
- Enabled a test that uses fully random characters that was previously disabled
- Fixes a bug where the transpiler did not detect choice expressions that cuDF is not capable of compiling
- Updated the compatibility guide to remove out of date information and group related items together
- Updated some tests to use `\z` instead of `$`

Follow-on issues:

- https://github.com/NVIDIA/spark-rapids/issues/4532
- https://github.com/NVIDIA/spark-rapids/issues/4533

### CODE REVIEW

File: tests/src/test/scala/com/nvidia/spark/rapids/RegularExpressionTranspilerSuite.scala [hunk 1]
```diff
@@ -193,50 +197,19 @@ class RegularExpressionTranspilerSuite extends FunSuite with Arm {
     assert(transpiled === expected)
   }
 
-  test("transpile complex regex 1") {
```
Comment (andygrove) [inline tests/src/test/scala/com/nvidia/spark/rapids/RegularExpressionTranspilerSuite.scala:196]: This regex is not supported by cuDF but the transpiler previously did not detect this

File: docs/compatibility.md [hunk 2]
```diff
@@ -465,25 +465,22 @@ CPU when the RAPIDS Accelerator determines that a pattern is either unsupported
 
 Here are some examples of regular expression patterns that are not supported on the GPU and will fall back to the CPU.
 
+- Line anchor `^` is not supported in some contexts, such as when combined with a choice (`^|a`).
+- Line anchor `$`
+- String anchor `\Z`
+- String anchors `\z` is not supported by `regexp_replace`
```
Comment (jlowe) [inline docs/compatibility.md:471]:
```suggestion
- String anchor `\z` is not supported by `regexp_replace`
```

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala [hunk 3]
```diff
@@ -441,16 +441,7 @@ class CudfRegexTranspiler(replace: Boolean) {
...
-            RegexRepetition(
-              RegexCharacterClass(negated = false,
-                characters = ListBuffer(RegexChar('\r'))),
-              SimpleQuantifier('?')),
-            RegexRepetition(
-              RegexCharacterClass(negated = false,
-                characters = ListBuffer(RegexChar('\n'))),
-              SimpleQuantifier('?')),
-            RegexChar('$')))
+          throw new RegexUnsupportedException("line anchor $ is not supported")
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala:444]: Should this have a link to the issue as is done below for the same case?

Comment (andygrove) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala:445]: Yes, and the other case was actually a bug .. an escaped `$` should have been dealt with as a literal `$` and not an anchor.

**Review by jlowe — APPROVED** [review]

**Comment by jlowe** [comment]
build

**Comment by andygrove** [comment]
build

**Review by jlowe — APPROVED** [review]

**Comment by jlowe** [comment]
build

**Pull request merged**